xdsl/dialects/riscv.py

```diff
@@ -3866,6 +3858,59 @@ class GetFloatRegisterOp(GetAnyRegisterOperation[FloatRegisterType]):
     name = "riscv.get_float_register"
 
 
+@irdl_op_definition
+class ParallelMovOp(IRDLOperation):
+    name = "riscv.parallel_mov"
+    inputs = var_operand_def(RISCVRegisterType)
+    outputs = var_result_def(RISCVRegisterType)
+    free_registers = opt_prop_def(ArrayAttr[RISCVRegisterType])
+
+    assembly_format = "$inputs attr-dict `:` functional-type($inputs, $outputs)"
+    irdl_options = [ParsePropInAttrDict()]
+
+    def __init__(
+        self,
+        inputs: Sequence[SSAValue],
+        outputs: Sequence[RISCVRegisterType],
+        free_registers: Sequence[RISCVRegisterType] | None = None,
+    ):
+        if free_registers is None:
+            properties = None
+        else:
+            properties = {"free_registers": ArrayAttr(free_registers)}
+
+        super().__init__(
+            operands=(inputs,),
+            result_types=(outputs,),
+            properties=properties,
+        )
+
+    def verify_(self) -> None:
+        if len(self.inputs) != len(self.outputs):
+            raise VerifyException(
+                "Input count must match output count. "
+                f"Num inputs: {len(self.inputs)}, Num outputs: {len(self.outputs)}"
+            )
+
+        input_types = self.inputs.types
+        output_types = self.outputs.types
+
+        # Check type of register type matches for input and output
+        for input_type, output_type in zip(input_types, output_types, strict=True):
+            if type(input_type) is not type(output_type):
+                raise VerifyException("Input type must match output type.")
+
+        # Check outputs are distinct if allocated and not ZERO
+        unallocated_registers = [Registers.UNALLOCATED_INT, Registers.UNALLOCATED_FLOAT]
```

Comment:
same comment here, this restricts the register types to a closed set, but it could also be checked with `output_type.is_allocated`